File: 1000-9999/2572.py
'''
2572번
보드게임
'''

import sys
input=sys.stdin.readline
sys.setrecursionlimit(10**9)

n=int(input())
color=list(input().split())
m, k=map(int, input().split())

road=[]; graph={}
for _ in range(m+1):
    road.append([0 for _ in range(m+1)])

for _ in range(k):
    a, b, c=input().split()
    a=int(a); b=int(b)
    road[a][b]=c
    road[b][a]=c
    if a not in graph: graph[a]=[]
    if b not in graph: graph[b]=[]
    graph[a].append(b)
    graph[b].append(a)

dp=[]
for _ in range(n):
    dp.append([-1 for _ in range(m+1)])

def solve(i, c):
    global dp, n
    if c==n: return 0
    if dp[c][i]!=-1: return dp[c][i]

    for g in graph[i]:
        s=0
        if road[i][g]==color[c]: s=10
        dp[c][i]=max(dp[c][i], solve(g, c+1)+s)
    return dp[c][i]

print(solve(1, 0))

File: 1000-9999/2572.cpp
//2572번 보드게임

#include <bits/stdc++.h>
using namespace std;

int n;
char color[1001];
int road[501][501];
vector<int> graph[501];
int dp[1001][501];

int solve(int i, int c){
    if(c==n) return 0;
    if(dp[c][i]!=-1) return dp[c][i];
    dp[c][i]=0;
    for(auto g:graph[i]){
        int s=0;
        if(road[i][g]==color[c]) s=10;
        dp[c][i]=max(dp[c][i], solve(g, c+1)+s);
    }
    return dp[c][i];
}

int main(void){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    int m, k, a, b;
    char c;
    cin>>n;
    for(int i=0; i<n; i++) cin>>color[i];
    cin>>m>>k;
    for(int i=0; i<k; i++){
        cin>>a>>b>>c;
        graph[a].push_back(b);
        graph[b].push_back(a);
        road[a][b]=c;
        road[b][a]=c;
    }

    for(int i=0; i<=1000; i++){
        for(int j=0; j<=500; j++){
            dp[i][j]=-1;
        }
    }

    cout<<solve(1, 0)<<"\n";
}